rfcs/20230418-constant-tensor-cache-capacity-control/README.md

```diff
@@ -0,0 +1,294 @@
+size reaches the capacity.
+
+A simple policy is: not cache the new coming tensors and not evict the cached
+tensors, when the cache size reach the capacity. We can called this policy as
+`NO_EVICT` simply. However, this policy may be not optimal sometimes. For
+example, generating different constant tensor may have different cost (like
+consumed time, memory, ...), users may prioritize evicting those low cost
+constant tensors and caching those high cost ones.
+
+Based on above considerations, we could extend the above capacity control API to
+allowing to set/get cache management policy. The API users could try different
+policies for their own scenarios and select the best one.
+
+```cpp
+typedef enum  {
+    NO_EVICT,
+    COST,
+    ... // Others
+} policy_t;
+
+void set_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, int size, policy_t policy);
+void get_constant_tensor_cache_capacity(dnnl::engine::kind eng_kind, int* size, policy_t * policy);
+```
+
+The environment variable value could be improved to `engine_kind:size:policy` as below:
+
+```shell
+export ONEDNN_CONSTANT_TENSOR_CACHE_CAPACITY=cpu:10240:NO_EVICT;gpu:2048:NO_EVICT
+```
+
+### Unified cache class design and backend API
+
+The key and value design of unified constant tensor cache:
+
+- Cache key: The unified cache key can be composed of two parts: backend
+  specific key and backend id. The backend specific key can be defined in each
+  backends, to achieve more flexibility. The backend id is used to avoid
+  repeated keys in different backends.
+- Cache value: The unified cache value can be a base buffer class like following,
+  and each backend can subclass it.
+
+```cpp
+class constant_buffer_t {
+public:
+    using malloc_func_t = void *(*)(size_t, const dnnl::engine &,
+            const graph::allocator_t *);
+    using free_func_t = void (*)(
+            void *, const dnnl::engine &, const graph::allocator_t *);
+
+    // Backends should provide these malloc and free function handles
+    constant_buffer_t(size_t size, const dnnl::engine &eng,
+            const graph::allocator_t *alc, malloc_func_t malloc_func,
+            free_func_t free_func)
+        : size_(size)
+        , eng_(eng)
+        , alc_(alc)
+        , malloc_func_(malloc_func)
+        , free_func_(free_func) {
+        data_ = malloc_func_(size, eng, alc);
+        const_cast<allocator_t *>(alc)->retain();
+    }
+
+    virtual ~constant_buffer_t() {
+        free_func_(data_, eng_, alc_);
+        const_cast<allocator_t *>(alc_)->release();
+    };
+
+    // Disable assignment and copy
+    constant_buffer_t(const constant_buffer_t &) = delete;
+    constant_buffer_t(constant_buffer_t &&) = delete;
+    constant_buffer_t &operator=(const constant_buffer_t &) = delete;
+    constant_buffer_t &operator=(constant_buffer_t &&) = delete;
+
+    template <typename T>
+    T *data() const {
+        return static_cast<T *>(data_);
+    }
+
+    size_t size() const { return size_; }
```

Comment:
Shall we have an additional virtual function to notify the backend that the buffer is evicted from cache? It would be very useful for us, since we are doing internal reference counting in the buffer in graph compiler runtime. It can be like `virtual void notify_evict() {}` with default implementation left empty. And we need `unified_constant_tensor_cache_t` to call `notify_evict` when `remove_if_exist`. Is this ok to you?

Reply:
Please note that `notify_evict` is unlike the destructor of `constant_buffer_t`. In `remove_if_exist`, if we remove the shared ptr of `remove_if_exist` in your internal cache, the destructor may not be called, because there may be other references.

Reply:
If I understand correctly, you need the cache to mark if the tensor is in cache or not by the proposed `notify_evict` api? Can you elaborate how the tensor will be handled if it has been evicted from the cache? How long it will be alive in backend?

Reply:
> Can you elaborate how the tensor will be handled if it has been evicted from the cache?
As planned, we will have an internal refcount for each buffer. Entering the kernel will increase the count and leaving the kernel will deref the count. The count is initialized as 1, indicating that it is in the cache. We also need the cache manager to deref the count if the buffer is no longer in the cache. When the refcount reaches 0, we can finally release the buffer safely.

> How long it will be alive in backend?
It will be freed when
 * the buffer is evicted, and if no kernel is referencing the buffer
 * or the buffer is evicted and the kernel referencing the buffer exits